src/serac/numerics/functional/detail/hexahedron_H1.inl

```diff
@@ -73,5 +96,385 @@ struct finite_element<Geometry::Hexahedron, H1<p, c> > {
+
+    static constexpr bool apply_weights = true;
+    static constexpr auto B             = calculate_B<apply_weights, q>();
+    static constexpr auto G             = calculate_G<apply_weights, q>();
+
+    cache_type<q> cache{};
+
+    for (int j = 0; j < ntrial; j++) {
+      for (int i = 0; i < c; i++) {
+        s_buffer_type source;
+        f_buffer_type flux;
+
+        for (int qz = 0; qz < q; qz++) {
+          for (int qy = 0; qy < q; qy++) {
+            for (int qx = 0; qx < q; qx++) {
+              int Q              = (qz * q + qy) * q + qx;
+              source(qz, qy, qx) = reinterpret_cast<const double*>(&get<SOURCE>(qf_output[Q]))[i * ntrial + j];
+              for (int k = 0; k < dim; k++) {
+                flux(k, qz, qy, qx) =
+                    reinterpret_cast<const double*>(&get<FLUX>(qf_output[Q]))[(i * dim + k) * ntrial + j];
+              }
+            }
+          }
+        }
+
+        cache.A2[0] = contract<2, 0>(source, B) + contract<2, 0>(flux(0), G);
+        cache.A2[1] = contract<2, 0>(flux(1), B);
+        cache.A2[2] = contract<2, 0>(flux(2), B);
+
+        cache.A1[0] = contract<1, 0>(cache.A2[0], B) + contract<1, 0>(cache.A2[1], G);
+        cache.A1[1] = contract<1, 0>(cache.A2[2], B);
+
+        element_residual[j * step](i) += contract<0, 0>(cache.A1[0], B) + contract<0, 0>(cache.A1[1], G);
+      }
+    }
+  }
+
+#if defined(__CUDACC__)
+
+  template <int q>
+  static SERAC_DEVICE auto interpolate(const dof_type& X, const tensor<double, dim, dim>& J,
+                                       const TensorProductQuadratureRule<q>& rule, cache_type<q>& cache)
+  {
+    // we want to compute the following:
+    //
+    // X_q(u, v, w) := (B(u, i) * B(v, j) * B(w, k)) * X_e(i, j, k)
+    //
+    // where
+    //   X_q(u, v, w) are the quadrature-point values at position {u, v, w},
+    //   B(u, i) is the i^{th} 1D interpolation/differentiation (shape) function,
+    //           evaluated at the u^{th} 1D quadrature point, and
+    //   X_e(i, j, k) are the values at node {i, j, k} to be interpolated
+    //
+    // this algorithm carries out the above calculation in 3 steps:
+    //
+    // A1(dz, dy, qx)  := B(qx, dx) * X_e(dz, dy, dx)
+    // A2(dz, qy, qx)  := B(qy, dy) * A1(dz, dy, qx)
+    // X_q(qz, qy, qx) := B(qz, dz) * A2(dz, qy, qx)
+
+    int tidx = threadIdx.x % q;
+    int tidy = (threadIdx.x % (q * q)) / q;
+    int tidz = threadIdx.x / (q * q);
+
+    static constexpr auto points1D = GaussLegendreNodes<q>();
+    static constexpr auto B_       = [=]() {
+      tensor<double, q, n> B{};
+      for (int i = 0; i < q; i++) {
+        B[i] = GaussLobattoInterpolation<n>(points1D[i]);
+      }
+      return B;
+    }();
+
+    static constexpr auto G_ = [=]() {
+      tensor<double, q, n> G{};
+      for (int i = 0; i < q; i++) {
+        G[i] = GaussLobattoInterpolationDerivative<n>(points1D[i]);
+      }
+      return G;
+    }();
+
+    __shared__ tensor<double, q, n> B;

```

Comment:
@white238 , ideally these would be dynamically allocated. Do you know if umpire (or some other tool) can do this with a single CPU/CUDA implementation?